Disable self-registration by default

IntegralProgrammer opened the pull request:
As default settings should always be the safest settings, this PR causes the default behavior of Apache Sling to deny self-registration attempts. Additional users therefore must be added by logging in as `admin` and creating the new user account.

If the administrator wishes to enable self-registration, they can do so by logging in as `admin`, visiting `/system/console/configMgr`, editing `Apache Sling Create User` and enabling self-registration.

IntegralProgrammer committed:
Disable self-registration by default

sdumitriu reviewed (changes requested):
This does achieve the desired purpose, but it leaves the application "broken". As it is, trying to "request an account" apparently does nothing, clicking the submit button doesn't show any error message, it just... does nothing.

Possible fixes:
1. Remove the "request an account" button completely (easy fix)
2. Add some error handling for the `500` server response, displaying the returned message (`Sorry, registration of new users is not currently enabled. Please try again later.`)
3. Actually implement the "Request" aspect, instead of sending a "Create account" request that will create and activate the account, it will create a disabled account that will have to be manually activated by an admin. I think there may be already a setting for that.
4. Option 3 will still create an account, even if disabled. A step further would be to literally make it an "account request", where we just store the details in a table, and only when the admin approves such a request will a real account be created.

I think doing 1 for this PR should be enough as a first step.

IntegralProgrammer commented:
> This does achieve the desired purpose, but it leaves the application "broken". As it is, trying to "request an account" apparently does nothing, clicking the submit button doesn't show any error message, it just... does nothing.
> 
> Possible fixes:
> 
>     1. Remove the "request an account" button completely (easy fix)
> 
>     2. Add some error handling for the `500` server response, displaying the returned message (`Sorry, registration of new users is not currently enabled. Please try again later.`)
> 
>     3. Actually implement the "Request" aspect, instead of sending a "Create account" request that will create and activate the account, it will create a disabled account that will have to be manually activated by an admin. I think there may be already a setting for that.
> 
>     4. Option 3 will still create an account, even if disabled. A step further would be to literally make it an "account request", where we just store the details in a table, and only when the admin approves such a request will a real account be created.
> 
> 
> I think doing 1 for this PR should be enough as a first step.

I could do option 1, but how should the sign-up button be re-added if self-registration is enabled? I could make it an ExtensionPoint, but wouldn't that require reloading with a different runmode?

marta- commented:
> I could do option 1, but how should the sign-up button be re-added if self-registration is enabled? I could make it an ExtensionPoint, but wouldn't that require reloading with a different runmode?

Is there a url you can ping to determine if self registration is enabled? If yes, you can simply display the "Don't have an account? Register" part based on the response from that URL.

IntegralProgrammer committed:
Display an error dialog containing the reason why sign-up failed should such type of failure occur.

TODO: Fixup the styling of the error dialog

IntegralProgrammer committed:
Fix styling of error dialog

IntegralProgrammer committed:
Handle failure of JSON parsing of error description

IntegralProgrammer commented:
I have addressed the review comment from @sdumitriu by following suggested option 2.

This may be out of scope for this PR, but line 233 (`this.form.setFieldError ...`) causes an error in the browser console (`_this3.form is undefined`). Is that line still needed?  Given the fact that if self-registration is enabled and an attempt is made to create a user with an already taken username, this error will be displayed in the error dialog, I think it can be removed. Thoughts?

sdumitriu commented:
> This may be out of scope for this PR, but line 233 (`this.form.setFieldError ...`) causes an error in the browser console (`_this3.form is undefined`). Is that line still needed? Given the fact that if self-registration is enabled and an attempt is made to create a user with an already taken username, this error will be displayed in the error dialog, I think it can be removed. Thoughts?

Hm, not sure why that line is there. Could serve a purpose when the form is displayed inline following a save error? I'll take a deeper look.

sashaandjic commented:
This is the current default behavior when clicking on SIGN IN button.
![image](https://user-images.githubusercontent.com/23243244/98878692-98d55200-2451-11eb-99ed-56ede82bbded.png)

sdumitriu commented:
@IntegralProgrammer `this.form` used to be a ref to the Formik component, but between the old Formik version that we used to use and the new one, they change it from a class component to a functional component, and thus the support for getting a ref to it was broken. This means that:

- there's a JS error logged about setting the ref prop on Formik
- `this.form` stopped working since it was set through that ref

We can either:
- try to fix the ref, there are a few hits that offer workarounds for this, the simplest one being to use `innerRef` instead of `ref`
- remove this functionality since we already display a dialog saying that the username is taken

IntegralProgrammer commented:
> @IntegralProgrammer `this.form` used to be a ref to the Formik component, but between the old Formik version that we used to use and the new one, they change it from a class component to a functional component, and thus the support for getting a ref to it was broken. This means that:
> 
>     * there's a JS error logged about setting the ref prop on Formik
> 
>     * `this.form` stopped working since it was set through that ref
> 
> 
> We can either:
> 
>     * try to fix the ref, there are a few hits that offer workarounds for this, the simplest one being to use `innerRef` instead of `ref`
> 
>     * remove this functionality since we already display a dialog saying that the username is taken

I vote to remove this functionality (option 2). Any objections? @marta-

marta- commented:
> > We can either:
> > ```
> > * try to fix the ref, there are a few hits that offer workarounds for this, the simplest one being to use `innerRef` instead of `ref`
> > 
> > * remove this functionality since we already display a dialog saying that the username is taken
> > ```
> 
> I vote to remove this functionality (option 2). Any objections? @marta-

If `innerRef` instead of `ref` is a quick enough fix, I'd rather use that one. But if it turns out to be more to it than just that, ok to remove the functionality.

IntegralProgrammer committed:
Handle setting of error text on the sign-up form username field

IntegralProgrammer commented:
@marta- I have made this fix of changing `ref` to `innerRef` along with some other minor changes and the bug appears to be fixed. Please let me know if additional changes are required, and if not, then @sashaandjic can test.
